usr/sbin/p11sak/p11sak.c

```diff
@@ -113,20 +114,57 @@ static CK_RV get_pin(char **pin, size_t *pinlen) {
     }
 
     /* strip the carriage return since not part of pin. */
-    buff[nread - 1] = '\0';
-    memcpy(*pin, buff, nread);
+    user_input[nread - 1] = '\0';
+    memcpy(*pin, user_input, nread);
     /* don't include the terminating null in the pinlen */
     *pinlen = nread - 1;
 
-    done: if (buff)
-        free(buff);
+    done: if (user_input)
+        free(user_input);
+
+    return rc;
+}
+
+static CK_RV get_keyboard_input(char **user_input)
+{
+    int nread;
+    char *keyboard_input = NULL;
+    size_t buflen;
+    CK_RV rc = 0;
+
+    /* read the keyboard_input
+     * Note: getline will allocate memory for keyboard_input. free it when done.
+     */
+    nread = getline(&keyboard_input, &buflen, stdin);
```

Comment:
Why not directly pass user_input to getline? 

*user_input  = NULL;
nread = getline(user_input , &buflen, stdin);

Then you do not need to allocate a new buffer and copy the content.

Reply:
Reworked user_input to remove allocating new buffers & content copy.